src/target/riscv/riscv-013.c

```diff
@@ -2567,6 +2566,14 @@ static bool riscv013_is_halted(struct target *target)
 		LOG_ERROR("hart %d is unavailable", riscv_current_hartid(target));
 	if (get_field(dmstatus, DMI_DMSTATUS_ANYNONEXISTENT))
 		LOG_ERROR("hart %d doesn't exist", riscv_current_hartid(target));
+	if (get_field(dmstatus, DMI_DMSTATUS_ANYHAVERESET)) {
+		int hartid = riscv_current_hartid(target);
+		LOG_INFO("hart %d unexpectedly reset!", hartid);
+		/* TODO: Can we make this more obvious to eg. a gdb user? */
+		dmi_write(target, DMI_DMCONTROL,
```

Comment:
if this was the case, shouldn't you do something like replace breakpoints? Or halt? Or...?

Reply:
Halting would be reasonable if the hart was halted when it was reset. I'll add that. Whether breakpoints should be replaced is much more iffy. Upon reset the code could be completely different, and breakpoints completely meaningless. The DRAM controller might not be set up so memory might not even be accessible. Really the debugger (gdb/eclipse) should know a bit more and be able to deal with a reset, but unfortunately there isn't even an interface to tell them that the hart was reset.

Reply:
yeah, good point. I guess it makes sense if you were single-stepping through the code in GDB if you could at least get it to print a message through GDB then the user could re-add breakpoints as necessary. I guess if you wanted to get fancy you could read the BP registers and see if they were still set or not so that the GDB "breakpoints" command would give the right information, but I'm not sure if OpenOCD->GDB can "cancel" breakpoints in this way.